general/echo/kmdf/exe/src/main.rs

```diff
@@ -0,0 +1,489 @@
+static WRITER_TYPE: u32 = 2;
+static NUM_ASYNCH_IO: usize = 100;
+static BUFFER_SIZE: usize = 40 * 1024;
+
+// The below implementation is a bit of a compromise in trying to help new devs find a 1 to 1 mapping
+// to the original C sample app code versus a full proper Rust implementation
+fn main() -> Result<(), Box<dyn Error>> {
+    let argv: Vec<String> = env::args().collect();
+    let argc = argv.len();
+
+    {
+        let mut globals = GLOBAL_DATA.write()?;
+        if argc > 1 {
+            if argv[1] == "-Async" {
+                globals.g_perform_async_io = true;
+                if argc > 2 {
+                    globals.g_async_io_loops_num = argv[2].parse::<usize>()?;
+                    globals.g_limited_loops = true;
+                } else {
+                    globals.g_limited_loops = false;
+                }
+            } else {
+                eprintln!(
+                    r##"
+Usage:
+    Echoapp.exe         --- Send single write and read request synchronously
+    Echoapp.exe -Async  --- Send reads and writes asynchronously without terminating
+    Echoapp.exe -Async <number> --- Send <number> reads and writes asynchronously
+Exit the app anytime by pressing Ctrl-C
+"##
+                );
+                return Err(String::from("Invalid Args").into());
+            }
+        }
+    }
+
+    get_device_path(&GUID_DEVINTERFACE_ECHO)?;
+
+    let globals = GLOBAL_DATA.read()?;
+    println!("DevicePath: {}", globals.g_device_path);
+
+    let h_device: HANDLE;
+    unsafe {
+        let mut path_vec = globals.g_device_path.encode_utf16().collect::<Vec<_>>();
+        path_vec.push(0);
+        let path = path_vec.as_ptr();
+
+        h_device = CreateFileW(
+            path,
+            FILE_GENERIC_READ | FILE_GENERIC_WRITE,
+            FILE_SHARE_READ | FILE_SHARE_WRITE,
+            std::ptr::null(),
+            OPEN_EXISTING,
+            0,
+            0,
+        );
+
+        if h_device == INVALID_HANDLE_VALUE {
+            return Err(
+                String::from(format!("Failed to open device. Error {}", GetLastError())).into(),
+            );
+        }
+    }
+
+    println!("Opened device successfully");
+
+    if globals.g_perform_async_io {
+        println!("Starting AsyncIo");
+
+        let h =
+            thread::spawn(|| -> Result<(), Box<dyn Error + Send + Sync>> { async_io(READER_TYPE) });
+
+        match async_io(WRITER_TYPE) {
+            Err(e) => return Err(e),
+            _ => (),
+        };
+
+        h.join().unwrap().unwrap();
+    } else {
+        perform_write_read_test(h_device, 512)?;
+
+        perform_write_read_test(h_device, 30 * 1024)?;
+    }
+
+    Ok(())
+}
+
+fn create_pattern_buffer(length: u32) -> Result<Vec<u8>, Box<dyn Error>> {
+    let mut buf = Vec::<u8>::with_capacity(length as usize);
+
+    for i in 0..length {
+        buf.push(i as u8);
+    }
+
+    Ok(buf)
+}
+
+fn verify_pattern_buffer(buf: &Vec<u8>) -> Result<(), Box<dyn Error>> {
```

Comment:
```suggestion
fn verify_pattern_buffer(buf: &[u8]) -> Result<(), Box<dyn Error>> {
```

Passing via slice instead of Vec reference avoids an extra layer of pointer indirection

Reply:
Fixed